ex_5/hashizume/EX5_k.py

```diff
@@ -0,0 +1,69 @@
+import numpy as np
+import matplotlib.pyplot as plt
+import sys
+
+
+def minimax(data, c_size):
+    data_size, n_features = data.shape
+    init_codebook = np.zeros((c_size, n_features))
+    rng = np.random.default_rng()
+    cent = rng.choice(data)
+    init_codebook[0] = cent
+    d = np.linalg.norm(data - cent, axis=1)
+    for i in range(1, c_size):
+        cent = data[np.argmax(d)]
+        init_codebook[i] = cent
+        d += np.linalg.norm(data - cent, axis=1)
+    return init_codebook
+
+
+def kmeans(data, c_size, init_codebook):
+    data_size, n_features = data.shape
+    cluster = np.zeros(data_size)
+    new_codebooks = np.zeros((c_size, n_features))
+    codebooks = init_codebook
+    while 1:
+        for i in range(data_size):
+            distances = np.linalg.norm(data[i] - codebooks, axis=1)
+            cluster[i] = np.argsort(distances)[0]
```

Comment:
```suggestion
        # (c_size,data_size,n_features)<-(1,data_size,n_features)-(c_size,1,n_features)
        differences = data[np.newaxis,:,:] - codebooks[:,np.newaxis,:]
        # data 側は broadcast により省略可能
        # differences = data - codebooks[:,np.newaxis,:]
        # (c_size, data_size) <- (c_size, data_size, n_features)
        distances = np.linalg.norm(differences, axis=2)
        # (data_size) <- (c_size, data_size)
        cluster = np.argmin(distances, axis=0)
```
ここは ndarray の broadcast 機能を最大限活かして for 文を削るのがいいと思います. 演算対象の ndarray への次元追加には np.newaxis を使うと shape を直感的に捉えることができるのでおすすめです. [https://note.nkmk.me/python-numpy-newaxis/](url)
それから, 最小をとるインデックスが欲しいだけなら np.argsort より np.argmin の方が速いと思います.

Reply:
for文使わず書くの慣れてないのでアドバイスとても助かります！！
np.newaxis 知らなかったので勉強になりました！
いつもありがとうございます〜（ ;  ; ）